addons/snailmail/static/src/js/thread_widget.js

```diff
@@ -0,0 +1,224 @@
+                {text: _t("Cancel letter"), close: true, click: _.bind(message._cancelLetter, message)},
+                {text: _t("Close"), close: true},
+            ],
+        }).open();
+    },
+    /**
+     * @private
+     * @param {integer} messageID 
+     * @param {string} content 
+     */
+    _openGenericErrorDialog: function (messageID, content) {
+        var message = _.find(this._messages, function (message) {
+            return message.getID() === messageID;
+        });
+        new Dialog(this, {
+            size: 'medium',
+            title: _t("Failed letter"),
+            $content: $('<div>').html(content),
+            buttons: [
+                {text: _t("Cancel letter"), classes: 'btn-primary', close: true, click: _.bind(message._cancelLetter, message)},
+                {text: _t("Close"), close: true},
+            ],
+        }).open();
+    },
+    /**
+     * Render the popover when mouse-hovering on the mail icon of a message
+     * in the thread. There is at most one such popover at any given time.
+     *
+     * @private
+     * @param {mail.model.AbstractMessage[]} messages list of messages in the
+     *   rendered thread, for which popover on mouseover interaction is
+     *   permitted.
+     */
+    _renderMessageSnailmailPopover: function () {
+        var self = this;
+        if (this._messageSnailmailPopover) {
+            this._messageSnailmailPopover.popover('hide');
+        }
+        if (!this.$('.o_thread_snailmail_tooltip').length) {
+            return;
+        }
+        this._messageSnailmailPopover = this.$('.o_thread_snailmail_tooltip').popover({
+            html: true,
+            boundary: 'viewport',
+            placement: 'auto',
+            trigger: 'hover',
+            offset: '0, 1',
+            content: function () {
+                var messageID = $(this).data('message-id');
+                var message = _.find(self._messages, function (message) {
+                    return message.getID() === messageID;
+                });
+                return QWeb.render('snailmail.widget.Thread.Message.SnailmailTooltip', {
+                    status: message.getSnailmailStatus()
+                });
+            },
+        });
+    },
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     * @param {MouseEvent} event
+     */
+    _onClickCreditError: function (ev) {
+        var self = this;
+        var messageID = $(ev.currentTarget).data('message-id');
+        this._rpc({
+            model: 'iap.account',
+            method: 'get_credits_url',
+            args: ['snailmail'],
+        }).then(function (link) {
+            var content = '<p>The letter could not be sent due to insufficient credits on your IAP account.<br/>Click <a href="' + link + '" target="new">here</a> to buy a pack.</p>';
```

Comment:
turn into translatable string, use string interpolation (`'... %s ...'`) instead of string concatenation (`'...' + '...'`)